t/test.pl: Add ability to cancel an 'alarm' watchdog

khwilliamson opened the pull request:
A watchdog affects the entire rest of the file.  As more tests get added
to a test file, they could end up triggering the watchdog.  It's trivial
to cancel a watchdog that uses SIGALRM.  This commit adds the capability
to do so; checking that what gets cancelled actually was set via this
method.

The other watchdog methods remain uncancellable.  Patches welcome.
These were created so that one could set a watchdog for cases where we
are testing timers and alarms.  Those situations aren't as common.
Hence this commit allows cancellation in most cases, but not all.

khwilliamson committed:
t/test.pl: Outdent a wrongly indented line

khwilliamson committed:
pat_advanced.t: Cancel a watchdog timer

This timer had been skipped until
91d110e02db3b8905c4c48d5fdd76a8439de31f3.  But with it enabled, on
slower or heavily loaded systems the test file does not always complete.
Take advantage of the new ability to cancel watchdog timers to prevent
this situation.

Another possibility would to have been to move this test to be the final
one in the file.  But then effort has to be expended to keep it in that
position as new tests are added; and what if such tests also require a
watchdog?  It's simplest to have the watchdog affect only the single
test that needs it.

Yet another possibility is to make this test into fresh_perl.  But that
sort of interferes with what t/re/pat_advanced_thr.t does.

tonycoz commented:
The only problem I see here is systems without alarm() (Win32) won't have a watchdog in this particular case.

khwilliamson committed:
t/test.pl: Add ability to cancel an watchdog timer

A watchdog affects the entire rest of the file.  As more tests get added
to a test file, they could end up triggering the watchdog.  This commit
adds the capability to cancel a watchdog if the potentially problematic
test finishes in time.

tonycoz commented on t/test.pl:
```diff
@@ -1733,16 +1733,42 @@ sub warning_like {
 #        _AFTER_ the 'threads' module is loaded.
 sub watchdog ($;$)
 {
+    CORE::state $watchdog;
+    CORE::state $watchdog_thread;
```
From the top of test.pl:
```
# Do not rely on features found only in more modern Perls here, as some CPAN
# distributions copy this file and must operate on older Perls. Similarly, keep
# things, simple as this may be run under fairly broken circumstances.
```
I think that means we shouldn't be using state here.

khwilliamson replied:
Changed to a closure
